File: apps/api/src/export/document/vector-codec.ts
import type { VectorOp, VectorPage, VectorPathCommand } from './vector';

const BINARY_MAGIC = 0x31_56_45_54; // TVE1

const OP_FILL_PATH = 0;
const OP_STROKE_PATH = 1;

// const FILL_RULE_WINDING = 0;
const FILL_RULE_EVEN_ODD = 1;

// const LINE_CAP_BUTT = 0;
const LINE_CAP_ROUND = 1;
const LINE_CAP_SQUARE = 2;

// const LINE_JOIN_MITER = 0;
const LINE_JOIN_ROUND = 1;
const LINE_JOIN_BEVEL = 2;

const CMD_MOVE_TO = 0;
const CMD_LINE_TO = 1;
const CMD_QUAD_TO = 2;
const CMD_CUBIC_TO = 3;
const CMD_CLOSE_PATH = 4;

const parsePath = (view: DataView, offsetRef: { value: number }, count: number): VectorPathCommand[] => {
  const path: VectorPathCommand[] = [];

  for (let i = 0; i < count; i++) {
    const cmd = view.getUint8(offsetRef.value);
    offsetRef.value += 1;

    if (cmd === CMD_MOVE_TO) {
      const x = view.getFloat32(offsetRef.value, true);
      const y = view.getFloat32(offsetRef.value + 4, true);
      offsetRef.value += 8;
      path.push({ type: 'moveTo', x, y });
      continue;
    }

    if (cmd === CMD_LINE_TO) {
      const x = view.getFloat32(offsetRef.value, true);
      const y = view.getFloat32(offsetRef.value + 4, true);
      offsetRef.value += 8;
      path.push({ type: 'lineTo', x, y });
      continue;
    }

    if (cmd === CMD_QUAD_TO) {
      const cx = view.getFloat32(offsetRef.value, true);
      const cy = view.getFloat32(offsetRef.value + 4, true);
      const x = view.getFloat32(offsetRef.value + 8, true);
      const y = view.getFloat32(offsetRef.value + 12, true);
      offsetRef.value += 16;
      path.push({ type: 'quadTo', cx, cy, x, y });
      continue;
    }

    if (cmd === CMD_CUBIC_TO) {
      const c1x = view.getFloat32(offsetRef.value, true);
      const c1y = view.getFloat32(offsetRef.value + 4, true);
      const c2x = view.getFloat32(offsetRef.value + 8, true);
      const c2y = view.getFloat32(offsetRef.value + 12, true);
      const x = view.getFloat32(offsetRef.value + 16, true);
      const y = view.getFloat32(offsetRef.value + 20, true);
      offsetRef.value += 24;
      path.push({ type: 'cubicTo', c1x, c1y, c2x, c2y, x, y });
      continue;
    }

    if (cmd === CMD_CLOSE_PATH) {
      path.push({ type: 'closePath' });
      continue;
    }

    throw new Error(`Unknown vector path command: ${cmd}`);
  }

  return path;
};

export const parseVectorPageBinary = (bytes: Uint8Array): Omit<VectorPage, 'externalElements'> => {
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const offset = { value: 0 };

  const magic = view.getUint32(offset.value, true);
  offset.value += 4;
  if (magic !== BINARY_MAGIC) {
    throw new Error(`Unexpected vector binary magic: ${magic.toString(16)}`);
  }

  const width = view.getFloat32(offset.value, true);
  const height = view.getFloat32(offset.value + 4, true);
  offset.value += 8;

  const opCount = view.getUint32(offset.value, true);
  offset.value += 4;

  const ops: VectorOp[] = [];

  for (let i = 0; i < opCount; i++) {
    const opTag = view.getUint8(offset.value);
    offset.value += 1;

    const pathCount = view.getUint32(offset.value, true);
    offset.value += 4;
    const path = parsePath(view, offset, pathCount);

    const color: [number, number, number, number] = [
      view.getUint8(offset.value),
      view.getUint8(offset.value + 1),
      view.getUint8(offset.value + 2),
      view.getUint8(offset.value + 3),
    ];
    offset.value += 4;

    if (opTag === OP_FILL_PATH) {
      const fillRuleTag = view.getUint8(offset.value);
      offset.value += 1;
      ops.push({
        type: 'fillPath',
        path,
        color,
        fillRule: fillRuleTag === FILL_RULE_EVEN_ODD ? 'evenOdd' : 'winding',
      });
      continue;
    }

    if (opTag === OP_STROKE_PATH) {
      const width = view.getFloat32(offset.value, true);
      offset.value += 4;

      const lineCapTag = view.getUint8(offset.value);
      const lineJoinTag = view.getUint8(offset.value + 1);
      offset.value += 2;

      const lineCap = lineCapTag === LINE_CAP_ROUND ? 'round' : lineCapTag === LINE_CAP_SQUARE ? 'square' : 'butt';
      const lineJoin = lineJoinTag === LINE_JOIN_ROUND ? 'round' : lineJoinTag === LINE_JOIN_BEVEL ? 'bevel' : 'miter';

      ops.push({
        type: 'strokePath',
        path,
        color,
        width,
        lineCap,
        lineJoin,
      });
      continue;
    }

    throw new Error(`Unknown vector op tag: ${opTag}`);
  }

  return { width, height, ops };
};

File: crates/editor/src/render/vector_codec.rs
use super::outline::{
    VectorFillRule, VectorLineCap, VectorLineJoin, VectorOp, VectorPage, VectorPathCommand,
};

const MAGIC: u32 = 0x3156_4554; // TVE1

const OP_FILL_PATH: u8 = 0;
const OP_STROKE_PATH: u8 = 1;

const FILL_RULE_WINDING: u8 = 0;
const FILL_RULE_EVEN_ODD: u8 = 1;

const LINE_CAP_BUTT: u8 = 0;
const LINE_CAP_ROUND: u8 = 1;
const LINE_CAP_SQUARE: u8 = 2;

const LINE_JOIN_MITER: u8 = 0;
const LINE_JOIN_ROUND: u8 = 1;
const LINE_JOIN_BEVEL: u8 = 2;

const CMD_MOVE_TO: u8 = 0;
const CMD_LINE_TO: u8 = 1;
const CMD_QUAD_TO: u8 = 2;
const CMD_CUBIC_TO: u8 = 3;
const CMD_CLOSE_PATH: u8 = 4;

pub fn encode_vector_page(page: &VectorPage) -> Vec<u8> {
    let mut out = Vec::with_capacity(page.ops.len() * 128 + 32);
    write_u32(&mut out, MAGIC);
    write_f32(&mut out, page.width);
    write_f32(&mut out, page.height);
    write_u32(&mut out, page.ops.len() as u32);

    for op in &page.ops {
        match op {
            VectorOp::FillPath {
                path,
                color,
                fill_rule,
            } => {
                write_u8(&mut out, OP_FILL_PATH);
                write_u32(&mut out, path.len() as u32);
                write_path_commands(&mut out, path);
                out.extend_from_slice(color);
                write_u8(
                    &mut out,
                    match fill_rule {
                        VectorFillRule::Winding => FILL_RULE_WINDING,
                        VectorFillRule::EvenOdd => FILL_RULE_EVEN_ODD,
                    },
                );
            }
            VectorOp::StrokePath {
                path,
                color,
                width,
                line_cap,
                line_join,
            } => {
                write_u8(&mut out, OP_STROKE_PATH);
                write_u32(&mut out, path.len() as u32);
                write_path_commands(&mut out, path);
                out.extend_from_slice(color);
                write_f32(&mut out, *width);
                write_u8(
                    &mut out,
                    match line_cap {
                        VectorLineCap::Butt => LINE_CAP_BUTT,
                        VectorLineCap::Round => LINE_CAP_ROUND,
                        VectorLineCap::Square => LINE_CAP_SQUARE,
                    },
                );
                write_u8(
                    &mut out,
                    match line_join {
                        VectorLineJoin::Miter => LINE_JOIN_MITER,
                        VectorLineJoin::Round => LINE_JOIN_ROUND,
                        VectorLineJoin::Bevel => LINE_JOIN_BEVEL,
                    },
                );
            }
        }
    }

    out
}

fn write_u8(out: &mut Vec<u8>, value: u8) {
    out.push(value);
}

fn write_u32(out: &mut Vec<u8>, value: u32) {
    out.extend_from_slice(&value.to_le_bytes());
}

fn write_f32(out: &mut Vec<u8>, value: f32) {
    out.extend_from_slice(&value.to_le_bytes());
}

fn write_path_commands(out: &mut Vec<u8>, path: &[VectorPathCommand]) {
    for command in path {
        match command {
            VectorPathCommand::MoveTo { x, y } => {
                write_u8(out, CMD_MOVE_TO);
                write_f32(out, *x);
                write_f32(out, *y);
            }
            VectorPathCommand::LineTo { x, y } => {
                write_u8(out, CMD_LINE_TO);
                write_f32(out, *x);
                write_f32(out, *y);
            }
            VectorPathCommand::QuadTo { cx, cy, x, y } => {
                write_u8(out, CMD_QUAD_TO);
                write_f32(out, *cx);
                write_f32(out, *cy);
                write_f32(out, *x);
                write_f32(out, *y);
            }
            VectorPathCommand::CubicTo {
                c1x,
                c1y,
                c2x,
                c2y,
                x,
                y,
            } => {
                write_u8(out, CMD_CUBIC_TO);
                write_f32(out, *c1x);
                write_f32(out, *c1y);
                write_f32(out, *c2x);
                write_f32(out, *c2y);
                write_f32(out, *x);
                write_f32(out, *y);
            }
            VectorPathCommand::ClosePath => {
                write_u8(out, CMD_CLOSE_PATH);
            }
        }
    }
}
